drivers/spi/spi_sam0.c

```diff
@@ -0,0 +1,432 @@
+	}
+	regs->DATA.reg = 0;
+
+	while (len--) {
+		if (len != 0) {
+			while (!regs->INTFLAG.bit.DRE) {
+			}
+			regs->DATA.reg = 0;
+		}
+
+		while (!regs->INTFLAG.bit.RXC) {
+		}
+
+		*p++ = regs->DATA.reg;
+	}
+
+	/* Note that all transmits are complete and the RX buf is empty */
+}
+
+/* Fast path that writes and reads bufs of the same length */
+static void spi_sam0_fast_txrx(SercomSpi *regs, const struct spi_buf *tx_buf,
+			       struct spi_buf *rx_buf)
+{
+	const u8_t *psrc = tx_buf->buf;
+	u8_t *p = rx_buf->buf;
+	size_t len = rx_buf->len;
+
+	while (regs->INTFLAG.bit.RXC) {
+		(void)regs->DATA.reg;
+	}
+
+	if (len <= 0) {
+		return;
+	}
+
+	/* Load the first outgoing byte */
+	while (!regs->INTFLAG.bit.DRE) {
+	}
+	regs->DATA.reg = *psrc++;
+
+	while (len--) {
+		if (len != 0) {
+			while (!regs->INTFLAG.bit.DRE) {
+			}
+			regs->DATA.reg = *psrc++;
+		}
+
+		while (!regs->INTFLAG.bit.RXC) {
+		}
+
+		*p++ = regs->DATA.reg;
+	}
+
+	/* Note that all transmits are complete and the RX buf is empty */
+}
+
+/* Finish any ongoing writes and drop any remaining read data */
+static void spi_sam0_finish(SercomSpi *regs)
+{
+	while (!regs->INTFLAG.bit.TXC) {
+	}
+	while (regs->INTFLAG.bit.RXC) {
+		(void)regs->DATA.reg;
+	}
+}
+
+/* Fast path where every overlapping tx and rx buffer is the same length */
+static void spi_sam0_fast_transceive(struct spi_config *config,
+				     const struct spi_buf *tx_bufs,
+				     size_t tx_count, struct spi_buf *rx_bufs,
+				     size_t rx_count)
+{
+	const struct spi_sam0_config *cfg = config->dev->config->config_info;
+	SercomSpi *regs = cfg->regs;
+
+	while (tx_count != 0 && rx_count != 0) {
+		spi_sam0_fast_txrx(regs, tx_bufs, rx_bufs);
+		tx_bufs++;
+		tx_count--;
+		rx_bufs++;
+		rx_count--;
+	}
+
+	for (; tx_count != 0; tx_count--) {
+		spi_sam0_fast_tx(regs, tx_bufs++);
+	}
+
+	for (; rx_count != 0; rx_count--) {
+		spi_sam0_fast_rx(regs, rx_bufs++);
+	}
+
+	spi_sam0_finish(regs);
+}
+
+/* Returns true if the request is suitable for the fast
+ * path. Specifically, the bufs are a sequence of:
+ *
+ * - Zero or more RX and TX buf pairs where each is the same length.
+ * - Zero or more trailing RX only bufs
+ * - Zero or more trailing TX only bufs
+ */
+static bool spi_sam0_is_regular(const struct spi_buf *tx_bufs,
+				size_t tx_count, struct spi_buf *rx_bufs,
+				size_t rx_count)
+{
+	while (tx_count != 0 && rx_count != 0) {
+		if (tx_bufs->len != rx_bufs->len) {
+			return false;
+		}
+
+		tx_bufs++;
+		tx_count--;
+		rx_bufs++;
+		rx_count--;
+	}
+
+	return true;
+}
+
+static int spi_sam0_transceive(struct spi_config *config,
+			       const struct spi_buf *tx_bufs, size_t tx_count,
+			       struct spi_buf *rx_bufs, size_t rx_count)
+{
+	const struct spi_sam0_config *cfg = config->dev->config->config_info;
+	struct spi_sam0_data *data = config->dev->driver_data;
+	SercomSpi *regs = cfg->regs;
+	int err;
+
+	spi_context_lock(&data->ctx, false, NULL);
+
+	err = spi_sam0_configure(config);
+
```

Comment:
you can remove that empty line as you are testing the previous assignment  in the if below.

Reply:
Done.